src/coreclr/jit/copyprop.cpp

```diff
@@ -389,7 +391,8 @@ bool Compiler::optBlockCopyProp(BasicBlock* block, LclNumToLiveDefsMap* curSsaNa
             {
                 optCopyPropPushDef(tree, lclDefNode, curSsaName);
             }
-            else if (tree->OperIs(GT_LCL_VAR, GT_LCL_FLD) && tree->AsLclVarCommon()->HasSsaName())
+            else if (tree->OperIs(GT_LCL_VAR, GT_LCL_FLD) && tree->AsLclVarCommon()->HasSsaName() &&
+                     tree->gtVNPair.BothDefined())
```

Comment:
I would recommend removing this check. `optCopyProp` doesn't actually depends on the VNs being present for trees it examines, so this is effectively the same as removing the assert.

Reply:
`optCopyProp` does `assert(tree->gtVNPair.BothDefined())` which is why I added this.

But it's true that `optCopyProp` does not actually use the `tree->gtVNPair` : it looks up the VN in the `lvPerSsaData` for the SSA def of the local. Presumably this assert is a partial expression of the expectation that if a local has an SSA number, then its SSA def should have the same VN as the SSA use (I believe that should be true?). It would therefore seem odd to ignore the fact that a lclvar use has no VN but its SSA does, and just use the def. That seems to indicate something is odd/wrong, perhaps due to morph, like what happened here.

Of course, I could remove the assert from `optCopyProp`, too.

Reply:
> Presumably this assert is a partial expression of the expectation that if a local has an SSA number, then its SSA def should have the same VN as the SSA use (I believe that should be true?).

Here is how this situation came to be: `optCopyProp` used to use the tree VN, but was changed to use the definition VN in #65250, with the main motivation being `LCL_FLD` handling. I left the tree VN assert as it seemed valuable on its own (indeed, as you say, to catch missing VNs - we are generally sorely lacking in validating this aspect of morph).

Reply:
So it seems like I could either:
1. Remove both the `tree->gtVNPair.BothDefined()` check here and `assert(tree->gtVNPair.BothDefined())` in `optCopyProp` (so `optCopyProp` will no longer check for "missing" VNs), or
2. Leave it as I've written it -- with more testing/checking than maybe should be required, or
3. Something else? (e.g., asserting for and fixing "missing VNs" -- but that seems out of scope for this fix)

Do you have a preference?

Reply:
My preference would be to check for missing VNs and fix them (leave the assert without the check, or even migrate it to the general loop in `optBlockCopyProp`), but barring that removing the assert (`1`) seems better.

Reply:
I think the SSA checker should also be able to spot missing VNs, though not perhaps in as timely a manner.

Reply:
> I think the SSA checker should also be able to spot missing VNs, though not perhaps in as timely a manner.

Well, `fgDebugCheckSsa()` already has this comment: `* Verify VNs on uses match the VN on the def` :-)